### ISSUE TITLE
Maya Tool - Adding/Removing nodes to the Solver causes UI slow updates

### ISSUE THREAD

**Issue by david-cattermole** [david-cattermole/mayaMatchMoveSolver#18]
When adding or removing many nodes with many frames, the solver UI is especially slow.

This is caused by the "solver steps" needing to compile down to individual keyword arguments, for each frame. This loop in Python is much slower than anticipated when it was written. 

We should:
1) Diagnose the slow-down of the Python loop
2) Refactor the feature and push the loop into code that can execute faster.
3) Or, if all else fails as a short term fix, we can provide an "update" button the user must click to update the UI.

We also need to investigate the possible problems caused by using Maya callbacks to update the UI perhaps we can intelligently process the updates once, rather than each time a node is added.

**Comment by david-cattermole** [comment]
This issue is related to #44.

After doing some profiling most of the slowness seems to be caused 50% from compiling the Solver UI "Solver Steps" into "Solvers", and another 50% from calling `Collection.is_valid()` (which compiles the collection under the hood).

However another source of slowdown is multiple callbacks (one for each changed attribute) triggering updates to the UI. This UI update is fairly fast and is not noticable for small scenes or small frame ranges, however longer frame ranges and multiple attribute causes slow downs.

If there are less frames in the solver steps, then this problem is less apparent.

I will attempt to address this problem by:
- reducing the need to update the UI for each attribute change and instead calling the UI only once after a batch of attributes are locked/unlocked.
- refactor the Collection compiling logic, for performance.

To test the performance, use the "Toggle Bundle Lock" tool in a Collection with many attributes added to the UI.

**Comment by david-cattermole** [comment]
A third option has been used to temporarily fix the slow-down effect.

When adding components to the Solver UI the the triggered callbacks needed to update the UI. Each update caused a re-compile of the Solver Steps. Since this re-compile was actually unneeded, it has been removed until needed later. The result has been to speed up the UI considerably, however the true underlying problem still exists and should be fixed. 

The end date on this issue can be delayed for another release.

**Comment by david-cattermole** [comment]
This issue can be delayed until another release.

A performance problem was fixed in the Solver UI which caused any edits to the UI to re-compile the "Solver Steps". This re-compile is unneeded and has been removed resulting in better UI performance.

**Comment by david-cattermole** [comment]
In the latest commits, the performance of adding/removing attributes has been improved. 

Any new performance issues with adding/removing should addressed in a new issue.
